charm/ksql/lib/charms/layer/ksql.py

```diff
@@ -0,0 +1,177 @@
+# Licensed to the Apache Software Foundation (ASF) under one or more
+# contributor license agreements.  See the NOTICE file distributed with
+# this work for additional information regarding copyright ownership.
+# The ASF licenses this file to You under the Apache License, Version 2.0
+# (the "License"); you may not use this file except in compliance with
+# the License.  You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import ipaddress
+import os
+import netifaces
+import re
+import shutil
+import socket
+import yaml
+from pathlib import Path
+
+from charmhelpers.core import hookenv, host
+from charmhelpers.core.templating import render
+
+KSQL_PORT=8088
+KSQL_SNAP='ksql-server'
+KSQL_SERVICE='snap.{}.ksql-server.service'.format(KSQL_SNAP)
+# TODO: Change this to common, once the new ksql server snap is published
+KSQL_SNAP_COMMON='/var/snap/{}/current'.format(KSQL_SNAP)
+KSQL_CLUSTER_ID='production'
+
+class Ksql(object):
+    def open_ports(self):
+        hookenv.open_port(KSQL_PORT)
+    
+    def close_ports(self):
+        hookenv.close_port(KSQL_PORT)
+
+    def configure(self, kafka_units, network_interface=None):
+        kafka = []
+        for unit in kafka_units:
+            ip = resolve_private_address(unit['host'])
+            kafka.append('{}:{}'.format(ip, unit['port']))
+        kafka.sort()
+        kafka_connect = ','.join(kafka)
+
+        ip = get_ip_for_interface(network_interface) if network_interface else hookenv.unit_private_ip()
+        
+        context = {
+            'bootstrap_servers': kafka_connect,
+            'listener_addr': ':'.join(['0.0.0.0', str(KSQL_PORT)]),
+            'keystore_password': _read_keystore_password(),
+            'ca_keystore': os.path.join(
+                KSQL_SNAP_COMMON,
+                'etc',
+                "ksql.client.jks"
+            ),
+            'client_keystore': os.path.join(
+                KSQL_SNAP_COMMON,
+                'etc',
+                "ksql.client.truststore.jks"
+            ),
+            'ksql_cluster_id': KSQL_CLUSTER_ID
+        }
+
+        render(
+            source='ksql-server.properties',
+            target=os.path.join(KSQL_SNAP_COMMON, 'etc', 'ksql-server.properties'),
+            owner="root",
+            perms=0o644,
+            context=context
+        )
+
+        render(
+            source='log4j.properties',
+            target=os.path.join(KSQL_SNAP_COMMON, 'etc', 'log4j.properties'),
+            owner="root",
+            perms=0o644,
+            context={}
```

Comment:
This seems kind of weird, to render a template without interpolating anything.

Reply:
We kind of have to, context is required.

Reply:
If there's nothing to interpolate, we could just `shutil.copyfile`. It might still be worth rendering it as a template if we want to configure something in log4j though.. like log level.